orderer/common/cluster/clusterservice.go

```diff
@@ -0,0 +1,262 @@
+	Dispatcher                       Dispatcher
+	Logger                           *flogging.FabricLogger
+	StepLogger                       *flogging.FabricLogger
+	MinimumExpirationWarningInterval time.Duration
+	CertExpWarningThreshold          time.Duration
+	MembershipByChannel              map[string]map[uint64]*msp.SerializedIdentity
+	Lock                             sync.Mutex
+	AuthVerifier                     AuthRequestVerifier
+}
+
+type AuthRequestSignature struct {
+	Version   int64
+	Timestamp string
+	FromId    string
+	ToId      string
+	Channel   string
+}
+
+// Step passes an implementation-specific message to another cluster member.
+func (s *ClusterService) Step(stream orderer.ClusterNodeService_StepServer) error {
+	s.StreamCountReporter.Increment()
+	defer s.StreamCountReporter.Decrement()
+
+	addr := util.ExtractRemoteAddress(stream.Context())
+	commonName := commonNameFromContext(stream.Context())
+	exp := s.initializeExpirationCheck(stream, addr, commonName)
+	s.Logger.Debugf("Connection from %s(%s)", commonName, addr)
+	defer s.Logger.Debugf("Closing connection from %s(%s)", commonName, addr)
+
+	// expect auth message is the first msg on the stream
+	request, err := stream.Recv()
+	if err == io.EOF {
+		s.Logger.Debugf("%s(%s) disconnected well before establishing the stream", commonName, addr)
+		return nil
+	}
+	if err != nil {
+		s.Logger.Warningf("Stream read from %s failed: %v", addr, err)
+		return err
+	}
+	channel, err := s.AuthVerifier.VerifyAuthRequest(stream, request, GetTLSSessionBinding, VerifySignature)
+	if err != nil {
+		s.Logger.Warnf("service authentication failed with error: %v", err)
+		return err
+	}
+
+	for {
+		err := s.handleMessage(stream, addr, exp, channel)
+		if err == io.EOF {
+			s.Logger.Debugf("%s(%s) disconnected", commonName, addr)
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		// Else, no error occurred, so we continue to the next iteration
+	}
+}
+
+type (
+	TLSSessionBindingGetter func(grpc.Stream, []byte) ([]byte, error)
+	SignatureVerifier       func([]byte, []byte, []byte) error
+)
+
+func (s *ClusterService) VerifyAuthRequest(stream orderer.ClusterNodeService_StepServer, request *orderer.ClusterNodeServiceStepRequest, tlsBindingGetter TLSSessionBindingGetter, signatureVerifier SignatureVerifier) (string, error) {
+	authReq := request.GetNodeAuthrequest()
+	if authReq == nil {
+		return "", status.Error(codes.Unauthenticated, "Invalid request object")
+	} else {
+		bindingFieldsHash := GetSessionBindingHash(authReq)
+
+		tlsBinding, err := tlsBindingGetter(stream, bindingFieldsHash)
+		if err != nil {
+			status.Errorf(codes.Unauthenticated, "Session binding error %v", err)
+			return authReq.Channel, err
+		}
+
+		if bytes.Equal(tlsBinding, authReq.SessionBinding) {
+			return authReq.Channel, status.Error(codes.Unauthenticated, "Session binding mismatch")
+		}
+
+		identity := s.MembershipByChannel[authReq.Channel][authReq.FromId]
+		if identity == nil {
```

Comment:
From hereafter, we should only return "access denied" and log the error, instead of telling a potential attacker why he was unable to authenticate to us.